internal/pubsub/gcppubsub/gcppubsub.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gcppubsub provides an implementation of pubsub that uses GCP
+// PubSub.
+package gcppubsub
+
+import (
+	"context"
+	"fmt"
+
+	rawgcppubsub "cloud.google.com/go/pubsub/apiv1"
+	"github.com/google/go-cloud/internal/pubsub"
+	"github.com/google/go-cloud/internal/pubsub/driver"
+	pubsubpb "google.golang.org/genproto/googleapis/pubsub/v1"
+)
+
+type topic struct {
+	path   string
+	client *rawgcppubsub.PublisherClient
+}
+
+// Close closes the connection to the API service. The user should invoke this
+// when the client is no longer required.
+func (t *topic) Close() error {
+	return t.client.Close()
+}
+
+// SendBatch publishes all the messages in dms. This method returns only after
+// all the messages are sent, an error occurs, or the context is cancelled.
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	var ms []*pubsubpb.PubsubMessage
+	for _, dm := range dms {
+		m := &pubsubpb.PubsubMessage{
+			Data:       dm.Body,
+			Attributes: dm.Metadata,
+		}
+		ms = append(ms, m)
+	}
+	req := &pubsubpb.PublishRequest{
+		Topic:    t.path,
+		Messages: ms,
+	}
+	_, err := t.client.Publish(ctx, req)
+	return err
+}
+
+type subscription struct {
+	client *rawgcppubsub.SubscriberClient
+	path   string
+}
+
+// ReceiveBatch returns a batch of messages that have queued up for the
+// subscription on the server. If no messages are available yet, it blocks
+// until there is at least one, or the context is done.
+func (s *subscription) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+	req := &pubsubpb.PullRequest{
+		Subscription:      s.path,
+		ReturnImmediately: false,
+	}
+	resp, err := s.client.Pull(ctx, req)
```

Comment:
You need a loop around this. Pull can still return empty-handed, even if `ReturnImmediately` is false.

Reply:
Okay. What error should I check for?

Reply:
Update: From our offline chat, the plan is to move the retry logic into the concrete API. See https://github.com/google/go-cloud/issues/722.